README.md
# Apache Cassandra Chef Cookbook

[![Build Status](https://travis-ci.org/michaelklishin/cassandra-chef-cookbook.svg?branch=master)](https://travis-ci.org/michaelklishin/cassandra-chef-cookbook)

This is a Chef cookbook for Apache Cassandra ([DataStax
Community Edition](http://www.datastax.com/products/community)) as
well as DataStax Enterprise.

It uses officially released packages and provides an Upstart service
script. It has fairly complete support for adjustment of Cassandra
configuration parameters using Chef node attributes.

It was originally created for CI and development environments and now supports cluster discovery using Chef search. **Feel free to contribute** what you find missing!

## Supported Chef Versions

This cookbook targets Chef 12 and later versions.

## Cookbook Dependencies

``` ruby
depends 'java'
depends 'ulimit'
depends 'apt'
depends 'yum'
depends 'ark'
```

## Cassandra Dependencies

Modern Cassandra versions require OracleJDK 8.


## Berkshelf

### Most Recent Release

``` ruby
cookbook 'cassandra-dse', '~> 4.5.0'
```

### From Git

``` ruby
cookbook 'cassandra-dse', github: 'michaelklishin/cassandra-chef-cookbook'
```


## Supported Apache Cassandra Version

This cookbook currently provides

 * Cassandra via tarballs
 * Cassandra (DataStax Community Edition) via apt and yum packages
 * DataStax Enterprise (DSE) via packages

## Supported OS Distributions

 * Ubuntu 12.04 through 17.101 via DataStax apt repo.
 * RHEL/CentOS via DataStax yum repo.
 * RHEL/CentOS/Amazon via tarball

## Support JDK Versions

Cassandra 2.x requires JDK 7+, later versions require Oracle JDK 8+.

## Recipes

The main recipe is `cassandra-dse::default` which together with the `node[:cassandra][:install_method]` attribute will be responsible for including the proper installation recipe and recipe `cassandra-dse::config` for configuring both `datastax` and `tarball` C* installation.

Two actual installation recipes are `cassandra-dse::tarball` and `cassandra-dse::datastax`. The former uses official tarball
and thus can be used to provision any specific version.

The latter uses DataStax repository via packages. You can install different versions (ex. `dsc20` for v2.0) available in the repository by altering `:package_name` attribute (`dsc20` by default).

Recently we have moved all the configuration resources to a separate recipe `cassandra-des::config`, which means recipes `cassandra-dse::tarball` and `cassandra-dse::datastax` are only responsible for C* installation.

>> Users with cookbook version `=<3.5.0` needs to update the `run_list`, in case of not using `cassandra-dse::default` recipe.

include_recipe `cassandra-dse` uses `cassandra-dse::datastax` as the default.

### DataStax Enterprise

You can also install the DataStax Enterprise edition by adding `node[:cassandra][:dse]` attributes according to the datastax.rb.
 * `node[:cassandra][:package_name]`: Override default value to 'dse-full'.
 * `node[:cassandra][:service_name]`: Override default value to 'dse'.

Unencrypted Credentials:
 * `node[:cassandra][:dse][:credentials][:username]`: Your username from Datastax website.
 * `node[:cassandra][:dse][:credentials][:password]`: Your password from Datastax website.

Encrypted Credentials:
 * `node[:cassandra][:dse][:credentials][:databag][:name]`: Databag name, i.e. the value 'cassandra' will reference to `/data_bags/cassandra`.
 * `node[:cassandra][:dse][:credentials][:databag][:item]`: Databag item, i.e. the value 'main' will reference to `/data_bags/cassandra/main.json`.
 * `node[:cassandra][:dse][:credentials][:databag][:entry]`: The field name in the databag item, in which the credetials are written. i.e. the data_bag:
```
{
  "id": "main",
  "entry": {
    "username": "%USERNAME%",
    "password": "%PASSWORD%"
  }
}
```

There are also recipes for DataStax opscenter installation (
`cassandra-dse::opscenter_agent_tarball`,
`cassandra-dse::opscenter_agent_datastax`, and
`cassandra-dse::opscenter_server` ) along with attributes available
for override (see below).

### JNA Support (for C* Versions Prior to 2.1.0)

The `node[:cassandra][:setup_jna]` attribute will install the jna.jar in the
`/usr/share/java/jna.jar`, and create a symbolic link to it on
`#{cassandra.lib\_dir}/jna.jar`, according to the [DataStax
documentation](http://www.datastax.com/documentation/cassandra/1.2/webhelp/cassandra/install/installJnaDeb.html).


## Node Attributes

Please note that the maintainers try to keep the list below up-to-date but it fairly often misses
some recently added attributes. Please refer to the [attributes files](https://github.com/michaelklishin/cassandra-chef-cookbook/tree/master/attributes) if an attribute you are looking for isn't listed.

### Core Attributes

 * `node[:cassandra][:install_method]` (default: datastax): The installation method to use (either 'datastax' or 'tarball').
 * `node[:cassandra][:config][:cluster_name]` (default: none): Name of the cluster to create. This is required.
 * `node[:cassandra][:version]` (default: a recent patch version): version to provision
 * `node[:cassandra][:tarball][:url]` and `node[:cassandra][:tarball][:sha256sum]` specify tarball URL and SHA256 check sum used by the `cassandra::tarball` recipe.
  * Setting `node[:cassandra][:tarball][:url]` to "auto" (default) will download the tarball of the specified version from the Apache repository.
 * `node[:cassandra][:setup_user]` (default: true): create user/group for Cassandra node process
 * `node[:cassandra][:setup_user_limits]` (default: true): setup Cassandra user limits
 * `node[:cassandra][:user]`: username Cassandra node process will use
 * `node[:cassandra][:group]`: groupname Cassandra node process will use
 * `node[:cassandra][:heap_new_size]` set JVM `-Xmn`. If set, `node[:cassandra][:max_heap_size]` must also be set; if nil, defaults to `min(100MB * num_cores, 1/4 * heap size)`
 * `node[:cassandra][:max_heap_size]` set JVM `-Xms` and `-Xmx`. If set, `node[:cassandra][:heap_new_size]` must also be set; if nil, defaults to `max(min(1/2 ram, 1024MB), min(1/4 ram, 8GB))`
 * `node[:cassandra][:installation_dir]` (default: `/usr/local/cassandra`): installation directory
 * `node[:cassandra][:root_dir]` (default: `/var/lib/cassandra`): data directory root
 * `node[:cassandra][:log_dir]` (default: `/var/log/cassandra`): log directory
 * `node[:cassandra][:tmp_dir]` (default: none): tmp directory. Be careful what you set this to, as the cassandra user will be given ownership of that directory.
 * `node[:cassandra][:local_jmx]` (default: true): bind JMX listener to localhost
 * `node[:cassandra][:jmx_port]` (default: 7199): port to listen for JMX
 * `node[:cassandra][:jmx_remote_rmi_port]` (default: $JMX_PORT): port for jmx remote method invocation. If using internode SSL, there is a bug requiring this to be different than `node[:cassandra][:jmx_port]`
 * `node[:cassandra][:jmx_remote_authenticate]` (default: false): turn on to require username/password for jmx operations including nodetool. To turn on requires `node[:cassandra][:local_jmx]` to be false
 * `node[:cassandra][:jmx][:user]` (default: cassandra): username for jmx authentication
 * `node[:cassandra][:jmx][:password]` (default: cassandra): password for jmx authentication.
 * `node[:cassandra][:notify_restart]` (default: false): notify Cassandra service restart upon resource update
  * Setting `node[:cassandra][:notify_restart]` to true will restart Cassandra service upon resource change
 * `node[:cassandra][:setup_jna]` (default: true): installs jna.jar
 * `node[:cassandra][:skip_jna]` (default: false): (2.1.0 and up only) removes jna.jar, adding '-Dcassandra.boot_without_jna=true' for low-memory C* installations
 * `node[:cassandra][:pid_dir]` (default: true): pid directory for Cassandra node process for `cassandra::tarball` recipe
 * `node[:cassandra][:dir_mode]` (default: 0755): default permission set for Cassandra node directory / files
 * `node[:cassandra][:service_action]` (default: [:enable, :start]): default service actions for the service
 * `node[:cassandra][:install_java]` (default: true): whether to run the open source java cookbook
 * `node[:cassandra][:cassandra_old_version_20]` (default: ): attribute used in cookbook to determine C* version older or newer than 2.1
 * `node[:cassandra][:log_config_files]` (default: calculated): log framework configuration files name array
 * `node[:cassandra][:xss]`  JVM per thread stack-size (-Xss option) (default: 256k).
 * `node[:cassandra][:jmx_server_hostname]` java.rmi.server.hostname option for JMX interface, necessary to set when you have problems connecting to JMX) (default: false)
 * `node[:cassandra][:heap_dump]` -XX:+HeapDumpOnOutOfMemoryError JVM parameter (default: true)
 * `node[:cassandra][:heap_dump_dir]` Directory where heap dumps will be placed (default: nil, which will use cwd)
 * `node[:cassandra][:vnodes]` enable vnodes. (default: true)

 For the complete set of supported attributes, please consult [the source](https://github.com/michaelklishin/cassandra-chef-cookbook/tree/master/attributes).

Attributes used to define JBOD functionality

* `default['cassandra']['jbod']['slices']` - defines the number of jbod slices while each represents data directory. By default disables with nil.
* `default['cassandra']['jbod']['dir_name_prefix']` - defines the data directory prefix
For example if you want to connect 4 EBS disks as a JBOD slices the names will be in the following format: data1,data2,data3,data4
cassandra.yaml.erb will generate automatically entry per data_dir location
Please note: this functionality is not creating volumes or directories. It takes care of configuration. You can use same parameters with AWS cookbook to create EBS volumes and map to directories.

Attributes for fine tuning CMS/ParNew, the GC algorithm recommended for Cassandra deployments:

 * `node[:cassandra][:gc_survivor_ratio]` -XX:SurvivorRatio JVM parameter (default: 8)
 * `node[:cassandra][:gc_max_tenuring_threshold]` -XX:MaxTenuringThreshold JVM parameter (default: 1)
 * `node[:cassandra][:gc_cms_initiating_occupancy_fraction]` -XX:CMSInitiatingOccupancyFraction JVM parameter (default: 75)

Descriptions for these JVM parameters can be found [here](http://www.oracle.com/technetwork/java/javase/tech/vmoptions-jsp-140102.html#PerformanceTuning) and [here](http://www.oracle.com/technetwork/java/javase/gc-tuning-6-140523.html#cms.starting_a_cycle).

Attributes for enabling G1 GC.

 * `node[:cassandra][:jvm][:g1]` (default: false)

Attributes for enabling GC detail/logging.

 * `node[:cassandra][:jvm][:gcdetail]` (default: false)

Attributes for fine tuning the G1 GC algorithm:

 * `node[:cassandra][:jvm][:g1_rset_updating_pause_time_percent]` (default: 10)
 * `node[:cassandra][:jvm][:g1_heap_region_size]` -XX:G1HeapRegionSize (default: 0)
 * `node[:cassandra][:jvm][:max_gc_pause_millis]` -XX:MaxGCPauseMillis (default: 200)
 * `node[:cassandra][:jvm][:heap_occupancy_threshold]` -XX:InitiatingHeapOccupancyPercent (default: 45)
 * `node[:cassandra][:jvm][:max_parallel_gc_threads]` This will set -XX:ParallelGCThreads to the number of cores on the machine (default: false)
 * `node[:cassandra][:jvm][:max_conc_gc_threads]` This will set -XX:ConcGCThreads to the number of cores on the machine (default: false)
 * `node[:cassandra][:jvm][:parallel_ref_proc]` -XX:ParallelRefProcEnabled (default: false)
 * `node[:cassandra][:jvm][:always_pre_touch]` -XX:AlwaysPreTouch (default: false)
 * `node[:cassandra][:jvm][:use_biased_locking]` -XX:UseBiasedLocking  (default: true)
 * `node[:cassandra][:jvm][:use_tlab]` -XX:UseTLAB (default: true)
 * `node[:cassandra][:jvm][:resize_tlab]` -XX:ResizeTLAB (default: true)

Oracle JVM 8 tuning parameters: [here](https://docs.oracle.com/javase/8/docs/technotes/guides/vm/gctuning/)

### Seed Discovery Attributes

 * `node[:cassandra][:seeds]` (default: `[node[:ipaddress]]`): an array of nodes this node will contact to discover cluster topology
 * `node[:cassandra][:seed_discovery][:use_chef_search]` (default: false): enabled seed discovery using Chef search
 * `node[:cassandra][:seed_discovery][:search_role]` (default: `"cassandra-seed"`): role to use in search query
 * `node[:cassandra][:seed_discovery][:search_query]` (default: uses `node[:cassandra][:seed_discovery][:search_role]`): allows
   for overriding the entire Chef search query
 * `node[:cassandra][:seed_discovery][:count]` (default: `3`): how many nodes to include into seed list. First N nodes are
   taken in the order Chef search returns them. IP addresses of the nodes are sorted lexographically.

### cassandra.yaml Attributes

 * `node[:cassandra][:config][:num_tokens]` set the desired number of tokens. (default: 256)
 * `node[:cassandra][:config][:listen_address]` (default: node[:ipaddress]): address clients will use to connect to the node
 * `node[:cassandra][:config][:broadcast_address]` (default: node IP address): address to broadcast to other Cassandra nodes
 * `node[:cassandra][:config][:rpc_address]` (default: 0.0.0.0): address to bind the RPC interface.  Leave blank to lookup IP from hostname.
 * `node[:cassandra][:config][:hinted_handoff_enabled]` see http://wiki.apache.org/cassandra/HintedHandoff (default: true)
 * `node[:cassandra][:config][:max_hint_window_in_ms]` The maximum amount of time a dead host will have hints generated (default: 10800000).
 * `node[:cassandra][:config][:hinted_handoff_throttle_in_kb]` throttle in KB's per second, per delivery thread (default: 1024)
 * `node[:cassandra][:config][:max_hints_delivery_threads]` Number of threads with which to deliver hints (default: 2)
 * `node[:cassandra][:config][:authenticator]` Authentication backend (default: org.apache.cassandra.auth.AllowAllAuthenticator)
 * `node[:cassandra][:config][:authorizer]` Authorization backend (default: org.apache.cassandra.auth.AllowAllAuthorizer)
 * `node[:cassandra][:config][:permissions_validity_in_ms]` Validity period for permissions cache, set to0 to disable (default: 2000)
 * `node[:cassandra][:config][:partitioner]` The partitioner to distribute keys across the cluster (default: org.apache.cassandra.dht.Murmur3Partitioner).
 * `node[:cassandra][:config][:disk_failure_policy]` policy for data disk failures: stop, best\_effort, or ignore (default: stop)
 * `node[:cassandra][:config][:key_cache_size_in_mb]` Maximum size of the key cache in memory. Set to 0 to disable, or "" for auto = (min(5% of Heap (in MB), 100MB)) (default: "", auto).
 * `node[:cassandra][:config][:key_cache_save_period]` Duration in seconds after which key cache is saved to saved\_caches\_directory. (default: 14400)
 * `node[:cassandra][:config][:row_cache_size_in_mb]` Maximum size of the row cache in memory, 0 to disable (default: 0)
 * `node[:cassandra][:config][:row_cache_save_period]` Duration in seconds after which row cache is saved to saved\_caches\_directory, 0 to disable cache save. (default: 0)
 * `node[:cassandra][:config][:row_cache_provider]` The provider for the row cache to use (default: SerializingCacheProvider)
 * `node[:cassandra][:config][:commitlog_sync]` periodic to ack writes immediately with periodic fsyncs, or batch to wait until fsync to ack writes (default: periodic)
 * `node[:cassandra][:config][:commitlog_sync_period_in_ms]` period for commitlog fsync when commitlog\_sync = periodic (default: 10000)
 * `node[:cassandra][:config][:commitlog_sync_batch_window_in_ms]` batch window for fsync when commitlog\_sync = batch (default: 50)
 * `node[:cassandra][:config][:commitlog_segment_size_in_mb]` Size of individual commitlog file segments (default: 32)
 * `node[:cassandra][:config][:commitlog_total_space_in_mb]` If space gets above this value (it will round up to the next nearest segment multiple), Cassandra will flush every dirty CF in the oldest segment and remove it. (default: 4096)
 * `node[:cassandra][:config][:concurrent_reads]` Should be set to 16 * drives (default: 32)
 * `node[:cassandra][:config][:concurrent_writes]` Should be set to 8 * cpu cores (default: 32)
 * `node[:cassandra][:config][:trickle_fsync]` Enable this to avoid sudden dirty buffer flushing from impacting read latencies.  Almost always a good idea on SSDs; not necessary on platters (default: false)
 * `node[:cassandra][:config][:trickle_fsync_interval_in_kb]` Interval for fsync when doing sequential writes (default: 10240)
 * `node[:cassandra][:config][:storage_port]` TCP port, for commands and data (default: 7000)
 * `node[:cassandra][:config][:ssl_storage_port]` SSL port, unused unless enabled in encryption options (default: 7001)
 * `node[:cassandra][:config][:listen_address]` Address to bind for communication with other nodes. Leave blank to lookup IP from hostname. 0.0.0.0 is always wrong. (default: node[:ipaddress]).
 * `node[:cassandra][:config][:broadcast_address]` Address to broadcast to other Cassandra nodes.  If '', will use listen\_address (default: '')
 * `node[:cassandra][:config][:start_native_transport]` Whether to start the native transport server (default: true)
 * `node[:cassandra][:config][:native_transport_port]` Port for the CQL native transport to listen for clients on (default: 9042)
 * `node[:cassandra][:config][:start_rpc]` Whether to start the Thrift RPC server (default: true)
 * `node[:cassandra][:config][:rpc_port]` Port for Thrift RPC server to listen for clients on (default: 9160)
 * `node[:cassandra][:config][:rpc_keepalive]` Enable keepalive on RPC connections (default: true)
 * `node[:cassandra][:config][:rpc_server_type]` sync for one thread per connection; hsha for "half synchronous, half asynchronous" (default: sync)
 * `node[:cassandra][:config][:thrift_framed_transport_size_in_mb]` Frame size for Thrift (maximum field length) (default: 15)
 * `node[:cassandra][:config][:thrift_max_message_length_in_mb]` Max length of a Thrift message, including all fields and internal Thrift overhead (default: 16)
 * `node[:cassandra][:config][:incremental_backups]` Enable hardlinks in backups/ for each sstable flushed or streamed locally. Removing these links is the operator's responsibility (default: false)
 * `node[:cassandra][:config][:snapshot_before_compaction]` Take a snapshot before each compaction (default: false)
 * `node[:cassandra][:config][:auto_snapshot]` Take a snapshot before keyspace truncation or dropping of column families.  If you set this value to false, you will lose data on truncation or drop (default: true)
 * `node[:cassandra][:config][:column_index_size_in_kb]` Add column indexes to a row after its contents reach this size (default: 64)
 * `node[:cassandra][:config][:compaction_throughput_mb_per_sec]` Throttle compaction to this total system throughput. Generally should be 16-32 times data insertion rate (default: 16)

ChangeLog.md
## Changes Between 4.5.0 and 4.6.0 (unreleased)

No changes yet.


## Changes Between 4.4.0 and 4.5.0 (25-12-2017)

This release contains **potentially breaking changes**.

### Chef 13-compatible Attribute Syntax

The cookbook now uses the standardized node attribute syntax.

GitHub issue: [#365](https://github.com/michaelklishin/cassandra-chef-cookbook/pull/365)


## More Supported Cassandra Versions

Supported 3.x versions now include `3.10.x` and `3.11.x`.

Contributed by Yaniv Marom Nachumi.

GitHub issue: [#367](https://github.com/michaelklishin/cassandra-chef-cookbook/pull/367)


### jamm Installation Now Requires the User to Opt-in

**This is a potentially breaking change**.

`cassandra.setup_jamm` must be set to `true`, similarly to `cassandra.setup_priam` and `cassandra.setup_jna`.

Contributed by Romain Gerard.

GitHub issue: [#363](https://github.com/michaelklishin/cassandra-chef-cookbook/pull/363).


### Test Suite Fixes

Contributed by Brian Skarda.

GitHub issue: [#361](https://github.com/michaelklishin/cassandra-chef-cookbook/pull/361).


### Add Missing Directories to DSE Recipe

The commitlog and saved_caches directories are added to the datastax installation method.

Contributed by Brian Skarda.

GitHub issue: [#360](https://github.com/michaelklishin/cassandra-chef-cookbook/pull/360).

## Changes Between 4.3.0 and 4.4.0 (13-07-2017)

This release contains **potentially breaking changes**.

### `cassandra-env.sh` and JVM Options Synced with Cassandra 3

This is a **potentially breaking change**.

GitHub issue: [#328](https://github.com/michaelklishin/cassandra-chef-cookbook/pull/328).

Contributed by William Dauchy.

### Configurable DSE Repo URI

GitHub issue: [#346](https://github.com/michaelklishin/cassandra-chef-cookbook/pull/346)

Contributed by Eyal Zekaria.

### Delayed systemd Service Notifications

GitHub issue: [#353](https://github.com/michaelklishin/cassandra-chef-cookbook/pull/353)

Contributed by Ignasi Barrera.

### Restart on-failure Condition in systemd Service Unit

GitHub issue: [#342](https://github.com/michaelklishin/cassandra-chef-cookbook/pull/342)

Contributed by Victor Demonchy.

### Support for JVM Agents

GitHub issue: [#348](https://github.com/michaelklishin/cassandra-chef-cookbook/pull/348)

Contributed by @mihasya.

### Pre-created Hints Directory

Hints directory will be created if configured.

GitHub issue: [#340](https://github.com/michaelklishin/cassandra-chef-cookbook/pull/340)

Contributed by Ryan Scheidter.

### Support for More Service Management Tools

GitHub issue: [#341](https://github.com/michaelklishin/cassandra-chef-cookbook/pull/341)

Contributed by @yanivmn.

### More Cassandra Versions Covered

GitHub issue: [#347](https://github.com/michaelklishin/cassandra-chef-cookbook/pull/347)

Contributed by @mihasya.

### Spec Suite Improvements

GitHub issues: [#334](https://github.com/michaelklishin/cassandra-chef-cookbook/pull/334),
               [#339](https://github.com/michaelklishin/cassandra-chef-cookbook/pull/339)

Contributed by Grant Ridder and Virender Khatri.

### Rubocop Fixes

GitHub issue: [#344](https://github.com/michaelklishin/cassandra-chef-cookbook/pull/344)

Contributed by Romain Hardouin.



## Changes Between 4.2.0 and 4.3.0 (10-01-2017)

### systemd Support

Contributed by William Dauchy, Romain Gerard, Jean-Francois Weber-Marx.

### Update jamm Versions for Several Cassandra Series (e.g. 3.x)

Contributed by William Dauchy and Corentin Chary.

### Cassandra Version Changes Force Node Restarts

Thanks to a neat template generation trick by William Dauchy.

### Custom Sections in `opscenterd.conf`

Contributed by eyalzek.

### JMX Authentication Support

Contributed by Andrew Nolan.

### Make Sure that System Directory Permissions Are Unmodified

Contributed by Jean-Francois Weber-Marx.

### Bring Back Support for `native_transport_max_threads`

Contributed by Corentin Chary.

### Corrected `bin` Paths When Installing via Packages

Contributed by Barthelemy Vessemont.

### `cassandra_metrics` Are Now Configurable

Contrbuted by eyalzek.

### Ensure `source_dir` and `installation_dir` Minus the Leaf Directory Exist

Contributed by Michal Jankowski.

### Permissions of Optional JAR Files

...are now set to `0440`.

### Updated Package Checksums

Contributed by Michal Jankowski.

### Correctly Create `commitlog_dir` and `saved_caches_dir`

Contributed by Ryan Scheidter.



## Changes Between 4.1.0 and 4.2.0 (22-09-2016)

### More Reliable Upgrades

Contributed by Seth Rosenblum.

### Temp Directory Attributes

`cassandra.tmp_dir` is a new attribute that controls JNA and
JVM temporary directory (`java.io.tmpdir`) location.

Contributed by Jack Bracken.

### Jamm Version Updates for 3.x Releases

Contributed by Corentin Chary and Anthony Rabier.

### Hints Directory Not Set for C* 2.x

Hints directory is no longer set for C* 2.x as it's only
supported in 3.x.

Contributed by William Dauchy.

### More Reasonable Streaming Operation Socket Timeout

`cassandra.config.streaming_socket_timeout_in_ms` now defaults to 1 hour.

Contributed by Dimitris Verraros.

### Syslog Appender

Add syslog appender to log to remote servers.

Contributed by Andrew Nolan.

### More Reliable Data Directory Management

Contributed by Michael Saffitz.

### 3.x Package Naming Fixes

Contributed by Jason J. W. Williams.

### OpsCenter Agent [Chef] Environment Fixes

Contributed by Radek Wierzbicki.

### New 3.x Settings

Contributed by Otavio Fernandes.

### Java 8 JVM Tuning options

A number of tuning options have been added to the cookbook to provide more 
knobs to control performance in the JVM. This work is centered around the
use-case of Oracle JDK 8 with the G1 GC.
 
Contributed by Matthew Silvey.

## Changes Between 4.0.0 and 4.1.0 (Dec 28, 2015)

### DSE Compatibiilty Bug Fixes

For example, the `cassandra` package shouldn't be installed
when DSE is provisioned.

Contributed by Bill Warner and Dimitris Verraros.



attributes/default.rb
default['systemd']['units_dir'] = '/etc/systemd/system'

default['java']['jdk_version'] = '7'
default['java']['install_flavor'] = 'oracle'
default['java']['set_default'] = true
default['java']['oracle']['accept_oracle_download_terms'] = true

default['cassandra']['tarball_purge'] = false
default['cassandra']['install_method'] = 'datastax'
default['cassandra']['install_java'] = true
default['cassandra']['notify_restart'] = false
default['cassandra']['initial_token'] = ''
default['cassandra']['service_name'] = 'cassandra'
default['cassandra']['user'] = 'cassandra'
default['cassandra']['group'] = 'cassandra'
default['cassandra']['setup_user'] = true
default['cassandra']['setup_user_limits'] = true
default['cassandra']['user_home'] = nil
default['cassandra']['system_user'] = true
default['cassandra']['version'] = '2.2.0'
default['cassandra']['use_systemd'] = false
default['cassandra']['use_initd'] = true

# jamm library was added in v0.8.0 and
# not required for later versions
default['cassandra']['setup_jamm'] = false

default['cassandra']['setup_priam'] = false
default['cassandra']['priam']['base_url'] = "http://search.maven.org/remotecontent?filepath=com/netflix/priam/priam-cass-extensions/#{node['cassandra']['version']}"
default['cassandra']['priam']['jar_name'] = "priam-cass-extensions-#{node['cassandra']['version']}.jar"
default['cassandra']['priam']['sha256sum'] = '9fde9a40dc5c538adee54f40fa9027cf3ebb7fd42e3592b3e6fdfe3f7aff81e1'

default['cassandra']['pid_dir'] = '/var/run/cassandra'
default['cassandra']['dir_mode'] = '0755'
default['cassandra']['service_action'] = %i[enable start]
default['cassandra']['jmx_port'] = 7199
default['cassandra']['local_jmx'] = true
default['cassandra']['jmx_remote_rmi_port'] = '$JMX_PORT'
default['cassandra']['jmx_remote_authenticate'] = false
default['cassandra']['jmx']['user'] = 'cassandra'
default['cassandra']['jmx']['password'] = 'cassandra'

default['cassandra']['limits']['memlock'] = 'unlimited'
default['cassandra']['limits']['nofile'] = 48_000
default['cassandra']['limits']['nproc'] = 'unlimited'

default['cassandra']['templates_cookbook'] = 'cassandra-dse'

default['cassandra']['root_dir'] = '/var/lib/cassandra' # data/ subdir added to this root
default['cassandra']['log_dir'] = '/var/log/cassandra'
default['cassandra']['rootlogger'] = 'INFO,stdout,R'

# Seed node discovery
default['cassandra']['seeds'] = node['ipaddress']

default['cassandra']['seed_discovery']['use_chef_search'] = false
default['cassandra']['seed_discovery']['count'] = 3
default['cassandra']['seed_discovery']['search_role'] = 'cassandra-seed'
default['cassandra']['seed_discovery']['search_query'] = nil

default['cassandra']['jbod']['slices'] = nil
default['cassandra']['jbod']['dir_name_prefix'] = 'data'

default['cassandra']['logback']['file']['max_file_size'] = '20MB'
default['cassandra']['logback']['file']['max_index'] = 20
default['cassandra']['logback']['file']['min_index'] = 1
default['cassandra']['logback']['file']['pattern'] = '%-5level [%thread] %date{ISO8601} %F:%L - %msg%n'

default['cassandra']['logback']['debug']['enable'] = false
default['cassandra']['logback']['debug']['max_file_size'] = '20MB'
default['cassandra']['logback']['debug']['max_index'] = 20
default['cassandra']['logback']['debug']['min_index'] = 1
default['cassandra']['logback']['debug']['pattern'] = '%-5level [%thread] %date{ISO8601} %F:%L - %msg%n'

default['cassandra']['logback']['stdout']['enable'] = true
default['cassandra']['logback']['stdout']['pattern'] = '%-5level [%thread] %date{ISO8601} %F:%L - %msg%n'

default['cassandra']['logback']['syslog']['enable'] = false
default['cassandra']['logback']['syslog']['host'] = 'localhost'
default['cassandra']['logback']['syslog']['facility'] = 'USER'
default['cassandra']['logback']['syslog']['pattern'] = '%-5level [%thread] %F:%L - %msg%n'

default['cassandra']['logback']['override_loggers'] = {}

default['cassandra']['log4j'] = {}

data_dir = []
if !node['cassandra']['jbod']['slices'].nil?
  node['cassandra']['jbod']['slices'].times do |slice_number|
    data_dir << ::File.join(node['cassandra']['root_dir'], "#{node['cassandra']['jbod']['dir_name_prefix']}#{slice_number}")
  end
else
  data_dir << ::File.join(node['cassandra']['root_dir'], 'data')
end

default['cassandra']['data_dir'] = data_dir

default['cassandra']['max_heap_size'] = nil
default['cassandra']['heap_new_size'] = nil
default['cassandra']['xss'] = '256k'
default['cassandra']['vnodes'] = true
default['cassandra']['enable_assertions'] = true
default['cassandra']['internode_compression'] = 'all' # all, dc, none
default['cassandra']['jmx_server_hostname'] = false
default['cassandra']['metrics_reporter']['enabled'] = false
default['cassandra']['metrics_reporter']['name'] = 'metrics-graphite'
default['cassandra']['metrics_reporter']['jar_url'] = 'http://search.maven.org/remotecontent?filepath=com/yammer/metrics/metrics-graphite/2.2.0/metrics-graphite-2.2.0.jar'
default['cassandra']['metrics_reporter']['sha256sum'] = '6b4042aabf532229f8678b8dcd34e2215d94a683270898c162175b1b13d87de4'
default['cassandra']['metrics_reporter']['jar_name'] = 'metrics-graphite-2.2.0.jar'
default['cassandra']['metrics_reporter']['config'] = {} # should be a hash of relevant config

default['cassandra']['jamm']['version'] = jamm_version(node['cassandra']['version'])
default['cassandra']['jamm']['base_url'] = jamm_url(node['cassandra']['jamm']['version'])
default['cassandra']['jamm']['jar_name'] = "jamm-#{node['cassandra']['jamm']['version']}.jar"
default['cassandra']['jamm']['sha256sum'] = jamm_sha256sum(node['cassandra']['jamm']['version'])

# log configuration files
default['cassandra']['log_config_files'] = node['cassandra']['version'] =~ /^[0-1]|^2.0/ ? %w[log4j-server.properties] : %w[logback.xml logback-tools.xml]

# Heap Dump
default['cassandra']['heap_dump'] = true
default['cassandra']['heap_dump_dir'] = nil

# GC tuning options
default['cassandra']['jvm']['g1'] = false
default['cassandra']['jvm']['gcdetail'] = true

default['cassandra']['jvm']['max_gc_pause_millis'] = 200
default['cassandra']['jvm']['heap_occupancy_threshold'] = 45
default['cassandra']['jvm']['parallel_ref_proc'] = false
default['cassandra']['jvm']['max_parallel_gc_threads'] = nil
default['cassandra']['jvm']['max_conc_gc_threads'] = nil
default['cassandra']['jvm']['g1_rset_updating_pause_time_percent'] = 5
default['cassandra']['jvm']['g1_heap_region_size'] = 0
default['cassandra']['jvm']['always_pre_touch'] = false
default['cassandra']['jvm']['use_biased_locking'] = true
default['cassandra']['jvm']['use_tlab'] = true
default['cassandra']['jvm']['resize_tlab'] = true

default['cassandra']['jvm']['agentpath'] = nil
default['cassandra']['jvm']['misc_java_agents'] = []

default['cassandra']['jvm']['misc_jvm_options'] = []

default['cassandra']['gc_survivor_ratio'] = 8
default['cassandra']['gc_max_tenuring_threshold'] = 1
default['cassandra']['gc_cms_initiating_occupancy_fraction'] = 75

default['cassandra']['jna']['base_url'] = 'https://github.com/twall/jna/raw/4.0/dist'
default['cassandra']['jna']['jar_name'] = 'jna.jar'
default['cassandra']['jna']['sha256sum'] = 'dac270b6441ce24d93a96ddb6e8f93d8df099192738799a6f6fcfc2b2416ca19'

default['cassandra']['tarball']['url'] = 'auto'

default['cassandra']['opscenter']['version'] = nil
default['cassandra']['opscenter']['server']['package_name'] = 'opscenter'
default['cassandra']['opscenter']['server']['port'] = '8888'
default['cassandra']['opscenter']['server']['interface'] = '0.0.0.0'
default['cassandra']['opscenter']['server']['authentication'] = false

default['cassandra']['opscenter']['cassandra_metrics']['ignored_keyspaces'] = %w[system OpsCenter]
default['cassandra']['opscenter']['cassandra_metrics']['ignored_column_families'] = []
default['cassandra']['opscenter']['cassandra_metrics']['1min_ttl'] = 604800
default['cassandra']['opscenter']['cassandra_metrics']['5min_ttl'] = 2419200
default['cassandra']['opscenter']['cassandra_metrics']['2hr_ttl'] = 31536000

default['cassandra']['opscenter']['custom_configuration'] = {}

default['cassandra']['opscenter']['agent']['package_name'] = 'datastax-agent'
default['cassandra']['opscenter']['agent']['download_url'] = nil
default['cassandra']['opscenter']['agent']['checksum'] = nil
default['cassandra']['opscenter']['agent']['install_dir'] = '/opt'
default['cassandra']['opscenter']['agent']['install_folder_name'] = 'opscenter_agent'
default['cassandra']['opscenter']['agent']['binary_name'] = 'opscenter-agent'
# if nil, will use search to get IP by server role
default['cassandra']['opscenter']['agent']['server_host'] = nil
default['cassandra']['opscenter']['agent']['use_chef_search'] = true
default['cassandra']['opscenter']['agent']['server_role'] = 'opscenter_server'
default['cassandra']['opscenter']['agent']['use_ssl'] = false
default['cassandra']['opscenter']['agent']['conf_dir'] = '/var/lib/datastax-agent/conf'
